### PR TITLE
FloatingButtons: Create a 'hub' for B-R floating action buttons

### PR DESCRIPTION
**Pull request by spen**
This PR aims to move our floating action buttons in to a single container and styling in such a way as to allow them to flow naturally against each other (as opposed to us setting explicit absolute positions).

screen shot 2018-08-29 at 19 04 23

### To Test

Because this change set touches a number of existing buttons, testing is, unfortunately, not going to be particularly straight forward... 

### Testing Community Translator

- The easy way to test that this works visually is to go to `layout/community-translator/launcher.jsx` and set the `isEnabled` property to `true`.
- Refresh the page and you should see the globe icon right above `InlineHelp`.
- How does it look? It should be aligned to the centre of `InlineHelp`
- Hovering over it should still animate as before - sliding out to the left to reveal copy

### Testing HappyChat

- Make sure you, or somebody else is available on hud-staging.happychat.io (Note: This is the staging environment)
- Make sure you're using an account that is eligible for happychat (Biz / Personal)
- Click on the inline-help icon and click the 'contact us' button
- Start a chat and minimize it
  - You should see an orange button, again it should be centred above the other two items

### Testing Directly

The Directly service is a particularly difficult case to test, at least without making some changes to the code, this is because we don't control the creation of this element via react (a Directly script handles it instead).
Only 1 in 10 user accounts are eligible for Directly (at random) and only those using `en` (not including`en-gb`) will see it. Even then a conversation would need to be initiated in order to see the Directly widget in it's minimized form, which is what we're concerned with here...

While I have been testing, I've done so by doing the following
- Change `getSupportVariation` to always return the directly support type:

```js
export default function getSupportVariation( state ) {
  return SUPPORT_DIRECTLY; 
  ...
```

- Open inline help...
- Use the dev-tools to `find` `class="<div directly-rtm is-minimized is-hidden">` and remove the `is-hidden` class only (keep `is-minized`).
- Hover over that div to see highlighting in the DOM, like so:

screen shot 2018-08-29 at 15 38 58

- This is where we have to use our imagination... We can see that the element is clear of the environment badge and the inline help button. I think for now, this will have to do. I've marked this as a potential follow up task - to insert this element in to the floating buttons hub once it's been attached to the DOM, somehow.

Testing Jetpack-connect routes:
- See test plan of #26921 

### Follow up items

There are still 3 items that will hopefully make their way in to this new set up, but still have work to be done in order to get them there.

These are:
- [ ] Directly - as mentioned in the testing, Directly handle the insertion of the directly widget. We might want to explore options for adding this to the floating buttons container so that it abides by our styling (and thus doesn't need absolute positioning as it currently does).
- [ ] EnvironmentBadge - This isn't critical as it's dev only and currently not an issue thanks to some absolute positioning, but it's definitely a nice-to-have. This would mean that any future additions push the EnvironmentBadge and we won't need to manually change the positioning.
- [ ] BuildStatus - Same applies as in EnvironmentBadge.

### CODE REVIEW

File: client/layout/floating-actions/style.scss [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+	align-items: flex-end;
+}
+
+.floating-actions__vertical-bar {
+	display: flex;
+	flex-direction: column-reverse;
+	align-items: flex-end;
+}
+
+.floating-actions__item {
```
Comment (andfinally) [inline client/layout/floating-actions/style.scss]: It looks like we're not actually using this class.

File: client/layout/community-translator/launcher.jsx [hunk 2]
```diff
@@ -29,7 +29,7 @@ class TranslatorLauncher extends React.PureComponent {
 		infoDialogVisible: false,
 		firstActivation: true,
 		isActive: translator.isActivated(),
-		isEnabled: translator.isEnabled(),
+		isEnabled: true,
```
Comment (andfinally) [inline client/layout/community-translator/launcher.jsx]: Is the `translator.isEnabled()` call an old check we don't need any more?

**Comment by andfinally** [comment]
This is great - it's going to be so nice being able to get at the help button without deleting the environment badge! 

Translator, HappyChat and Directly all tested OK individually. My HappyChat didn't turn up in the staging environment, for some reason - it seemed like my messages were going to prod. 

At times in testing I got both Directly and the chat button, even with `getSupportVariation` returning `SUPPORT_HAPPYCHAT` and after hard refresh. I'm sure this something to do with tweaking `getSupportVariation` and Directly's JS that won't happen in real life - I'm not sure if it's worth bumping the Directly container up a bit in case.

image

When I was testing this on mobile layout on the Plans page, the ribbons on the plans were overlaid over the help button. Adding `z-index: 1` to the `floating-actions` container seems to fix it.

image

File: client/layout/community-translator/launcher.jsx (hunk 2 above)
Comment (spen) [reply client/layout/community-translator/launcher.jsx]: Ahh, this was meant to be chopped out. I just hardcoded it to make sure it'd show up each time :)
I'll remove it and put the code change in the test plan 👌

**Comment by spen** [comment]
> When I was testing this on mobile layout on the Plans page, the ribbons on the plans were overlaid over the help button. Adding z-index: 1 to the floating-actions container seems to fix it.

Very nicely caught - addressed in https://github.com/Automattic/wp-calypso/pull/26955/commits/eb77d68a727a713a40fc00043ce37c6902f4ecdd using our `z-index` scss util :)

> At times in testing I got both Directly and the chat button, even with getSupportVariation returning SUPPORT_HAPPYCHAT and after hard refresh. I'm sure this something to do with tweaking getSupportVariation and Directly's JS that won't happen in real life - I'm not sure if it's worth bumping the Directly container up a bit in case.

I'm fairly confident that this wouldn't happen in production and if it did I think it'd be a bug out of scope with these changes. The one issue I still have regarding directly right now, is that it doesn't abide by the positioning flow of the reset of these floating buttons and so sits awkwardly higher than other buttons (imaging the happychat button is not there in your screen shot). That said I don't want the Directly button to get in the way of merging these improvements - I'd rather scratch that down as follow up work.

File: client/layout/index.jsx [hunk 3]
```diff
@@ -98,22 +96,27 @@ const Layout = createReactClass( {
...
+			{ 'is-support-user': this.props.isSupportUser },
+			{ 'has-no-sidebar': ! this.props.hasSidebar },
+			{ 'has-chat': this.props.chatIsOpen },
+			{ 'has-no-masterbar': this.props.masterbarIsHidden }
+		);
+		const loadingClass = classnames( {
+			layout__loader: true,
+			'is-active': this.props.isLoading,
+		} );
+		const isJetpackConnectSection = 'jetpack-connect' !== this.props.section.name;
```
Comment (spen) [inline client/layout/index.jsx]: @tyxla, this is my take on the additions made in #26921, see [line 159](https://github.com/Automattic/wp-calypso/pull/26955/files#diff-e8ab61833a6d85e586cf195046ed6f15R159) also.

Does this still do as it should?

File: client/layout/floating-actions/style.scss [hunk 4]
```diff
@@ -0,0 +1,21 @@
+.floating-actions {
+	position: fixed;
+	padding: 10px;
+	right: 0;
+	bottom: 0;
+	display: flex;
+	flex-direction: row-reverse;
+	align-items: flex-end;
```
[the remark this answers is not in the document]
Comment (andfinally) [reply client/layout/floating-actions/style.scss]: (It turned out this layout was necessary to accommodate the environment build badge.)

File: client/components/happychat/style.scss [hunk 5]
```diff
@@ -105,6 +105,42 @@
 	}
 }
 
+.button.happychat__button {
+	z-index: z-index( 'root', '.floating-help' );
+	margin: 8px;
+
```
Comment (andfinally) [inline client/components/happychat/style.scss]: Is the line break here deliberate?

File: client/layout/community-translator/style.scss [hunk 6]
```diff
@@ -1,19 +1,27 @@
...
 
 	&.is-active {
 		background: $orange-jazzy;
 	}
 }
 
+.community-translator__container {
+	position: relative;
+	height: 48px;
+    width: 100%;
```
Comment (andfinally) [inline client/layout/community-translator/style.scss]: Some space indents here.

File: client/components/happychat/style.scss (hunk 5 above)
Comment (spen) [reply client/components/happychat/style.scss]: Another good spot, fixed up and rebased :)

File: client/layout/community-translator/style.scss (hunk 6 above)
Comment (spen) [reply client/layout/community-translator/style.scss]: Pesky spaces! Also fixed up and rebased!

**Comment by spen** [comment]
Thanks and good point @tyxla! I think I was tired ;) 
I've fixed this up by just inlining the logic. I think the var name itself may have gotten more confusing!

I *think* this should be ready for another look over @andfinally if you wouldn't mind 🙏

File: client/layout/index.jsx [hunk 7]
```diff
@@ -98,22 +96,26 @@ const Layout = createReactClass( {
...
+			{ 'has-chat': this.props.chatIsOpen },
+			{ 'has-no-masterbar': this.props.masterbarIsHidden }
+		);
+		const loadingClass = classnames( {
+			layout__loader: true,
+			'is-active': this.props.isLoading,
+		} );
+		const shouldShowInlineHelp =
+			( 'jetpack-connect' !== this.props.section.name || currentRoute === '/jetpack/new' ) &&
+			currentRoute !== '/log-in/jetpack';
```
[the remark this answers is not in the document]
Comment (spen) [reply client/layout/index.jsx]: I thought the exact same when reviewing the original source of this logic, see my comment here: https://github.com/Automattic/wp-calypso/pull/26921#discussion_r216161776

It's subtle, but significant.

[the remark this answers is not in the document]
Comment (tyxla) [reply client/layout/index.jsx]: I understand that can be confusing 😉  The idea here is that:

* `jetpack-connect` is a separate section.
* `/jetpack/new` is a route from the ⬆️ section.
* `/log-in/jetpack` is a route from another section.
* We'd like to disable inline help for the entire `jetpack-connect` section (with the following exception).
* We'd like to keep Inline Help for the `/jetpack/new` route.
* We'd like to disable Inline Help for the `/jetpack/log-in` route.

[the remark this answers is not in the document]
Comment (sirreal) [reply client/layout/index.jsx]: I had to think this through a few times when I first encountered it. It's important to know that `/log-in/jetpack` is not in the `jetpack-connect` section. I don't know how you arrived at `true && true`, but the evaluation isn't always that simple.

- Hide it if the user is in the `jetpack-connect` section. **Exception**: show it on the `/jetpack/new` route in this section.
- Show it if the user is _not in_ the `jetpack-connect` section. **Exception**: hide it on the `/log-in/jetpack` route which is not in the `jetpack-connect` section.


4 demonstrative cases:

```
// In the section, not on the exception route. Hide it.
const section = 'jetpack-connect', route = '/jetpack/connect';
// evaluation
( false || false ) && true -> false && true -> false

// In the section, on the exception route. Show it.
const section = 'jetpack-connect', route = '/jetpack/new';
// evaluation
( false || true ) && true -> true && true -> true

// Not in the section, not on the Jetpack log-in route. Show it
const section = 'log-in', route = '/log-in';
// evaluation
( true || false ) && true -> true && true -> true

// Not In the section, on the Jetpack log-in route. Hide it.
const section = 'log-in', route = '/log-in/jetpack';
// evaluation
( false || true ) && false -> true && false -> false
```

Note that some branches wouldn't even be evaluated due to short circuiting, but I've included them for clarity.

[the remark this answers is not in the document]
Comment (andfinally) [reply client/layout/index.jsx]: Thanks for the explanation @tyxla and @sirreal, very helpful!

**Pull request merged**